### PR TITLE
Adding back references tests.

### PR DESCRIPTION
**Pull request by macieksmuga**

### CODE REVIEW

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesPagingIT.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+                                                                       .build();
+        final ListReferenceBasesResponse resp = client.references.getReferenceBases(refId, req);
+        return resp.getSequence();
+    }
+
+    /**
+     * Check that we can request a sequence of bases from
+     * {@link org.ga4gh.ctk.transport.protocols.Client.References#getReferenceBases(String,
+     * ListReferenceBasesRequest)}
+     * using a size twice as large as the full sequence, and receive the full sequence.
```
Comment (calbach) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesPagingIT.java]: This doc does not appear to agree with the actual behavior.

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesPagingIT.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        final ListReferenceBasesRequest req = ListReferenceBasesRequest.newBuilder()
+                .setStart(sequenceStart).setEnd(sequenceEnd)
+                .build();
+        final ListReferenceBasesResponse resp = client.references.getReferenceBases(refId, req);
+        final String expectedSequence = resp.getSequence();
+        assertThat(expectedSequence).hasSize((int)sequenceEnd);
+
+        // walk through the bases by incrementing start and end
+        long offset = 0L;
+        for (char expectedChar : expectedSequence.toCharArray()) {
```
Comment (calbach) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesPagingIT.java]: 150 serial requests seems like overkill. Perhaps a smaller range? Or larger intervals?

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesPagingIT.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        final String fullSequence = getFullSequence();
+        checkChunkOfBases(TestData.REFERENCE_BRCA1_LENGTH, fullSequence);
+    }
+
+    /**
+     * Check that if we request a sequence of bases from
+     * {@link org.ga4gh.ctk.transport.protocols.Client.References#getReferenceBases(String,
+     * ListReferenceBasesRequest)}
+     * using a size larger than the full sequence, it fails with an exception with HTTP status
+     * "Requested Range Not Satisfiable."
```
Comment (calbach) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesPagingIT.java]: This seems like reasonable behavior, but is it part of the spec?

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesSearchIT.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+    /**
+     * Retrieve all {@link Reference} objects for the default {@link ReferenceSet}. Check
+     * that each returned {@link Reference} has a valid-looking MD5.  (In the future, perhaps this
+     * test should compute the MD5 and compare it to what's returned.)
+     *
+     * @throws AvroRemoteException if there's a communication problem or server exception ({@link GAException})
+     */
+    @Test
+    public void checkMD5ChecksumAppearanceOfAllReferences() throws AvroRemoteException {
```
Comment (calbach) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesSearchIT.java]: This doesn't appear to be related to ReferenceBases, which is the name of the test.

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsPagingIT.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        final List<ReferenceSet> listOfReferenceSets = Utils.getAllReferenceSets(client);
+        assertThat(listOfReferenceSets).isNotEmpty();
+
+        // we will remove ReferenceSets from this Set and assert at the end that we have zero
+        final Set<ReferenceSet> setOfReferenceSets = new HashSet<>(listOfReferenceSets);
+        assertThat(listOfReferenceSets).hasSize(setOfReferenceSets.size());
+
+        // page through the ReferenceSets using the same query parameters
+        String pageToken = null;
+        for (ReferenceSet ignored : listOfReferenceSets) {
```
Comment (calbach) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsPagingIT.java]: This pagination pattern is repeated in a few places. Consider the test output when the endpoint returns fewer results than expected. This will result in having us loop back around to the start and emitting confusing assertion failures.

Perhaps accumulate all results into a set and perform a set difference at the end. Or else, assert that pageToken is non-null if there are still more expected reference sets.

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsPagingIT.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+     */
+    @Test
+    public void checkPagingOneByOneThroughReferenceSets() throws AvroRemoteException {
+
+        // retrieve them all
+        final List<ReferenceSet> listOfReferenceSets = Utils.getAllReferenceSets(client);
+        assertThat(listOfReferenceSets).isNotEmpty();
+
+        // we will remove ReferenceSets from this Set and assert at the end that we have zero
+        final Set<ReferenceSet> setOfReferenceSets = new HashSet<>(listOfReferenceSets);
```
Comment (calbach) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsPagingIT.java]: I think naming this expectedReferenceSets would make things clearer.

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsPagingIT.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+    /**
+     * Check that we can page through the {@link ReferenceSet}s we receive from
+     * {@link Client.References#searchReferenceSets(SearchReferenceSetsRequest)}
+     * using an increment as large as the non-paged set of results.
+     *
+     * @throws AvroRemoteException if there's a communication problem or server exception ({@link GAException})
+     */
+    @Test
+    public void checkPagingByOneChunkThroughReferenceSets() throws AvroRemoteException {
+
```
Comment (jeromekelleher) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsPagingIT.java]: This is unlikely to fail in practise, but the server is not *required* to satisfy the pageSize requests. It's perfectly entitled to send back the results in pages of 1, if it so desires. This won't fail now because we only have a handful of objects, but if there was 1000 reference sets, then some servers would send back the results in chunks smaller than 1000. As such, I think this is a bit problematic and should possibly be removed.

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsPagingIT.java [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    /**
+     * Check that we can page through the {@link ReferenceSet}s we receive from
+     * {@link Client.References#searchReferenceSets(SearchReferenceSetsRequest)}
+     * using an increment twice as large as the non-paged set of results.
+     *
+     * @throws AvroRemoteException if there's a communication problem or server exception ({@link GAException})
+     */
+    @Test
+    public void checkPagingByOneTooLargeChunkThroughReferenceSets() throws AvroRemoteException {
```
Comment (jeromekelleher) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsPagingIT.java]: This is similar. All we can assert here is that we should get back all the results after paging. The sizes of the pages are up to the server, and anything > 0 is a valid value.

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsSearchIT.java [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    @Test
+    public void checkSearchingReferenceSetsReturnsWellFormed() throws AvroRemoteException {
+        final SearchReferenceSetsRequest req =
+                SearchReferenceSetsRequest.newBuilder().build();
+        final SearchReferenceSetsResponse resp = client.references.searchReferenceSets(req);
+
+        final List<ReferenceSet> sets = resp.getReferenceSets();
+
+        sets.stream().forEach(rs -> assertThat(rs.getId()).isNotNull());
+        sets.stream().forEach(rs -> assertThat(rs.getMd5checksum()).isNotNull());
```
Comment (jeromekelleher) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsSearchIT.java]: Should we be checking that this looks like an MD5 also?

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsSearchIT.java [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+     * @throws AvroRemoteException if there's a communication problem or server exception ({@link GAException})
+     */
+    @Test
+    public void checkReferenceSetFoundByAccession() throws AvroRemoteException {
+        final SearchReferenceSetsRequest req =
+                SearchReferenceSetsRequest.newBuilder()
+                        .setAccessions(TestData.REFERENCESET_ACCESSIONS)
+                        .build();
+        final SearchReferenceSetsResponse resp = client.references.searchReferenceSets(req);
+
```
Comment (jeromekelleher) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsSearchIT.java]: These 3 methods are almost identical. The checking ``resp`` bit should be factored out.

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferencesPagingIT.java [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+        final List<Reference> listOfReferences = Utils.getAllReferences(client, refSetId);
+        assertThat(listOfReferences).isNotEmpty();
+
+        // we will remove References from this Set and assert at the end that we have zero
+        final Set<Reference> setOfReferences = new HashSet<>(listOfReferences);
+        assertThat(listOfReferences).hasSize(setOfReferences.size());
+
+        // page through the References using the same query parameters
+        String pageToken = null;
+        for (Reference ignored : listOfReferences) {
```
Comment (jeromekelleher) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferencesPagingIT.java]: This seems quite low-level. Why not provide a method to do the paging automatically, which you can provide an optional page size (which I think @calbach suggests above too)? There's going to be an awful lot of code repetition if this is done for every object, on every search. I would say this test should be replaced with something like
```java
final Set<Reference> expectedReferences = Set(Utils.getAllReferences(client, refSetId));
final Set<Reference> foundReferences = Set(getReferences(client, refSetId, pageSize=1));
assertThat(expectedReferences == foundReferences);
```
(apologies for the mangled pseudocode, but you get the idea!)

The point is, we can't really every assert anything interesting about pageSizes other than they are > 0, so there's no point in checking these things.

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferencesSearchIT.java [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+                        .setReferenceSetId(referenceSetId)
+                        .build();
+        final SearchReferencesResponse resp = client.references.searchReferences(req);
+
+        final List<Reference> sets = resp.getReferences();
+
+        sets.stream().forEach(rs -> assertThat(rs.getId()).isNotNull());
+        sets.stream().forEach(rs -> assertThat(rs.getName()).isNotNull());
+        sets.stream().forEach(rs -> assertThat(rs.getLength()).isNotNull());
+        sets.stream().forEach(rs -> assertThat(rs.getMd5checksum()).isNotNull());
```
Comment (jeromekelleher) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferencesSearchIT.java]: Looks like MD5?

**Comment by jeromekelleher** [comment]
This looks great, thanks @hjellinek and @macieksmuga. The code is easy to read and the tests are very nice. I  agree with all the comments @calbach makes above.

I'm slightly worried about these paging tests though. I think we should aim to centralise the paging code and concentrate on testing the stream of objects we get back. This is the interesting bit.

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesPagingIT.java (hunk 1 above)
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesPagingIT.java]: You're right, I fixed the comment.

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesPagingIT.java (hunk 3 above)
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesPagingIT.java]: Not yet, but see ga4gh/schemas#430.

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsPagingIT.java (hunk 7 above)
Comment (macieksmuga) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsPagingIT.java]: @jeromekelleher to be precise, my understanding is that a compliant server should never send back _more_ objects per request than specified by `pageSize` (it can certainly send less).

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsPagingIT.java (hunk 6 above)
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsPagingIT.java]: It's now, or will shortly be, `setOfExpectedReferenceSets`.

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsPagingIT.java (hunk 5 above)
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsPagingIT.java]: I had indeed ignored that case; I plead ignorance.  Is this documented?  If not, please open an issue against the schemas.  You can even assign it to me! :-)

I have updated the `checkPagingOneByOneThroughReferences` and `checkPagingOneByOneThroughReferenceSets` methods similarly to what you describe.

Comment (calbach) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsPagingIT.java]: My comment only pertains to the behavior of this specific test case, perhaps I could have written it more clearly. The reason it would loop back through is because you would set `pageToken` to null, effectively sending the same original request again.

In the scenario I described, the server is misbehaving in that it returns fewer results than expected - but the test should expect this as a possible failure scenario and should be clear in its failure message.

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsPagingIT.java (hunk 7 above)
Comment (jeromekelleher) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsPagingIT.java]: Fair enough @macieksmuga --- we must have 0 < returnedPageSize <= requestedPageSize.

**Comment by jeromekelleher** [comment]
Any updates on this @hjellinek?

**Comment by macieksmuga** [comment]
With the pull of @hjellinek's latest code, are people's issues answered? Can we get the +1's to make this mergeable?

**Comment by calbach** [comment]
I still see a bunch of unaddressed comments here (mostly from @jeromekelleher), I also just noticed one more issue.

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesPagingIT.java [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+     *
+     * @param chunkSize        the sequence length we'll request
+     * @param expectedSequence the full base sequence we expect
+     * @throws AvroRemoteException if there's a communication problem or server exception
+     */
+    private void checkChunkOfBases(long chunkSize,
+                                   String expectedSequence)
+            throws AvroRemoteException {
+
+        final String refId = Utils.getValidReferenceId(client);
```
Comment (calbach) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesPagingIT.java]: This approach only works because there's exactly one `Reference` in the system. How I expected this to work ideally, is that a helper would perform a lookup by assembly ID to find your known reference set, then do a lookup by reference set ID + reference name to get a specific reference ID.

Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesPagingIT.java]: Good one.  There's a lot of room for new references-related tests, and the surest route to that would be to add more references to the compliance data set.  It's difficult to write tests with generality or confidence when you know there's only a single instance of the object under test.

Once there are multiple references, I expect things like `Utils.getValidReferenceId(client)` to go away in most tests in favor of the approach you outlined.

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsSearchIT.java (hunk 9 above)
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsSearchIT.java]: I had a test like that previously, and class Utils already includes a method called `looksLikeValidMd5`.  I can add a call to that, but I think it would be better to add a check that the MD5 value is correct.  It's a much stronger test.

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsPagingIT.java (hunk 7 above)
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsPagingIT.java]: I believe all the paging tests now check this.

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsSearchIT.java (hunk 10 above)
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceSetsSearchIT.java]: Yup.  My philosophy, adapted from Martin Fowler: write unoptimized code, then refactor when you see a pattern or redundancy.  Or when @jeromekelleher does. :-)

Fixed, thanks.

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferencesSearchIT.java (hunk 12 above)
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferencesSearchIT.java]: Can we not do _equals the correct MD5_, @macieksmuga?

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesPagingIT.java (hunk 13 above)
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesPagingIT.java]: I had missed this batch of @jeromekelleher's comments, but believe I have now addressed them.

Comment (calbach) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesPagingIT.java]: The test has a hidden dependency on there being exactly one reference, which I consider to be an issue now - not only later, once more references are added. Ideally this would be fixed before the test is submitted, but if not there should at least be a TODO here (I wouldn't block my +1 on it being fixed here).

**Comment by jeromekelleher** [comment]
OK, happy to +1 (once #116 is merged?). We could do with some more refactoring to take out the reliance of looping over pageTokens, but this is a good start.

**Comment by hjellinek** [comment]
Thanks, @jeromekelleher.  I don't know that there's anything blocking #116, so I hope this can happen soon.

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesPagingIT.java (hunk 13 above)
Comment (hjellinek) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesPagingIT.java]: Thanks, @calbach. I just opened issues #117 and #118.  Issues are less likely to be overlooked than a TODO in the source code.

**Comment by jeromekelleher** [comment]
I think you can just update this PR by pushing to it directly @hjellinek. Having PRs against PRs seems like overkill...

File: cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesSearchIT.java (hunk 4 above)
Comment (calbach) [inline cts-java/src/test/java/org/ga4gh/cts/api/references/ReferenceBasesSearchIT.java]: This is still unaddressed, but I won't block on it. I guess you could say the MD5 field of the references is technically related to the reference bases, though this seems like a bit of a stretch.

**Comment by calbach** [comment]
+1, though I second the need for an eventual second pass on pagination here. There are still several places where the logic is not correct, though is likely to pass on most servers (e.g. all the tests which check pageSize * 2). It may make sense to pair a PR of pagination cleanup/fixes with some additional pagination documentation in the schemas branch.

**Comment by jeromekelleher** [comment]
Also +1, and completely agree with @calbach's comments.

**Comment by jeromekelleher** [comment]
I think we're ready to merge here. One question though: are we going to start squashing commits on PRs like this? Here we have some fairly boring updates and merge noise which would be much cleaner as a single commit. I'd prefer a clean history, but I guess it doesn't matter so much here in the compliance tests since we're not going to be making bugfix releases on older versions (probably??) so there won't be much cherry picking going on.

**Comment by macieksmuga** [comment]
I'll be happy to squish the commits here and in the other parallel PR's

**Comment by jeromekelleher** [comment]
Can we squash please @macieksmuga? Ready to merge otherwise I think.

**Comment by macieksmuga** [comment]
Squashed commits & rebased, @jeromekelleher. Should be ready to merge.

**Comment by jeromekelleher** [comment]
OK, we've got the requisite +1-age and a nice clean commit history, so I'm merging. Thanks @macieksmuga and @hjellinek!
